src/characters/porter.ts

```diff
@@ -50,6 +50,9 @@ type PostRetrieveCFragsResult = {
     readonly retrieval_results: readonly {
       readonly cfrags: {
         readonly [address: string]: string;
+      },
+      readonly errors: {
+        readonly [key: string]: string | Record<string, unknown>;
```

Comment:
@derekpierre I'm going to assume that this is the structure we're going for:
https://github.com/nucypher/tdec/issues/57#issuecomment-1252694555
I'll be happy to pair program over this

Reply:
Indeed. Looks like you've already encountered the format here: https://github.com/nucypher/nucypher/pull/2973#issuecomment-1253896820.

For now, the retrieval "errors" in the JSON will have the format: 
"<ursula_checksum_address>: <string>"
See https://github.com/derekpierre/nucypher/blob/porter-error-handling/nucypher/utilities/porter/control/specifications/fields/retrieve.py#L51.

Please note that multiple retrieval kits can be passed to the /retrieve_cfrags endpoint (as an efficiency mechanism), and if so, the result of retrieve_cfrags will look like the following, with the order of cfrags and errors matching the order the retrieval kits were provided in (just some fake error data from a test):

```json
{
   "retrieval_results":[
      {
         "cfrags":{
            "0x91d3c58Fe14e04585B8Af6196A9D461CBa034524":"A62wV78ncZgheIXeIpsrqS4UCnoDQrzu5hjPhBFKFTZlA5lurNvsZmiDs1eQa4Rgncp4q2fjkBHJzHgSLxG1P2NF9/rH87xG3ne0eSPb9EbouwECYbGUz9nGPS2gVg6zs+kCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDm4kQZWHCfMqelhs/sCpTbdYjow685EYGUuh2rqmwn/sDjJACFjOK7x3IfMS8f5SOfzUyAr9PM62gj0ZZmrVrblIDDLyPUhXP+QHTSl1GYw0BbWwM9eRHMrCDYRyInZuSUgMDht5sSbNbYbqYn/eLNKPJXDbGa7Yi+M83ZojrOGpjAE5sDjPSUlDFPh5Ai/YMF7pl3y7bCZAq2QHe4X/l2743H+ErMAUp177vEpqIcZcPA0uzqOHDrutIcq6Rv1ulAYSPMAORkHnMNYrKVNkEFn9a4UcZ2C88gutVeu0B7rb1FTs=",
            "0x9cB2a48e029b210F7454ec268d65Cf2B7F0914f8":"At/8gmvFc4u4JRKQlTb2n7syDQkz5yTx+Roxq7XSmq8mAssN0VDU5tHoySJaSIAB1W3Ng5ok0STCO7DMj+c6VPZ/Luj/LkGtMphva+2Y7RxIDffDcL98ko7xDoFpleIpYC0Ca4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDSfmCRtNdsJlA6pitLiyGe88r59QiwG6/2qLeBjwAPfEDgBjOh2ioV1CuUoC6Djexa6oX9RihyG3PTAN82dZZOpwCvHEhOv4dhB4ABP+1liPLiNxs1ZsrUQm2ZFMhmVwl/o4C0nW9Qfb6xLzDhmSCMNN4MtubyB5iD0A2Gr+KjNPAt3U+pPHLYNg0DjcgSUtnfkuKORb02MfbKDK/nNTCI0++zoclCSP+9IDd4G69up1/PFJCs/DNPp3FovzSd36VfPgEE4cu+D4ziA6LoZNQjlIM6AnDTuIq06r1SUx/6Ldnv7c=",
            "0xDD579AF829306D32AC17A48b92e6111376160E7F":"A+kxMCg3KvIj2fWnyK6+EUUvH+rKgpZJffPoqJcPCYBkA6Db+o88OPVDhnwxvqSKO+rCYrX26l6S/5+PpzEVPCZfXRvyefKLfqgTXPB2ei9pHJca9dogdmiQvFtxwvHb0fMCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoC+BqjNsJFrvPHMIxaBKaZ+Q00GFuj2bNRlyRORi8/0mADv0Qwk82z9WtWa58W97mYN0fKI58HrbahQAbNO5VBoQYDFM5QuMKzaDoF49QUqm/G7YbYJK+JbsHn9dP6Rhgkdl4C/7AmA4wyUh7EGojqq5Bp7wvFIY7m5bOl3d2hSDLtWiY56tqj7USDpkSFvwTPRi4MpnKLKZfUsoyoOOSg5DSupXbsDQaJnLIZbtRDKHafXAqQLpx2sdTf3YT5racl/6aMaA9Bi8NknqYakPMotYPSan59Icm0u1f7sS8lR9PZx9o="
         },
         "errors":{
            
         }
      },
      {
         "cfrags":{
            "0x91d3c58Fe14e04585B8Af6196A9D461CBa034524":"A8+0Y3e1MG06b1Syqm03UpCMsNKw0GIx0a2d3LA1PFUOA0WBCPqLv3mGMPl/Rls/bwJ/P3UVgqJziLUvKT2hbgOX9/rH87xG3ne0eSPb9EbouwECYbGUz9nGPS2gVg6zs+kCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDVSvdk2CWqUz39VlIc+HkAZCqwAWwoW+WxRAE8Kf9DN0CGFYZdHemkYoHA+kRqSkn1Fo9Ge4BRRTdw/c57zYRSPwDDLyPUhXP+QHTSl1GYw0BbWwM9eRHMrCDYRyInZuSUgMDiUmwH2cTxMFh1fCoiPrkroe8CcoX6F4TmQjHkqR+iO2Ewq7+HWR/c/f/Xv4O4Su+Dze29wa2s9g64ZwRKKO5g+ErMAUp177vEpqIcZcPA0uzqOHDrutIcq6Rv1ulAYSPMAORkHnMNYrKVNkEFn9a4UcZ2C88gutVeu0B7rb1FTs=",
            "0x9cB2a48e029b210F7454ec268d65Cf2B7F0914f8":"A9iFkm9P0HT8ZxrRjibbG7VwJcSZeeKMA3EIXdAmg7cAA+GkadwIg9yyQAsb5WOTHUgxXAQqdZyHfenkAFDLAoAcLuj/LkGtMphva+2Y7RxIDffDcL98ko7xDoFpleIpYC0Ca4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDQ8+7L67Vuea0MrXHEOok2OAtodcg5MDxM+x+fA0Hs2cCD4nFz9cEFU1e8mIHYQ+B2ASM9K67skXkSovomTM2iMgCvHEhOv4dhB4ABP+1liPLiNxs1ZsrUQm2ZFMhmVwl/o4CvKqzq+7xSWbIJDceLvZx4JyNJ4xt2KXR5+X562/YTtfv9C6FyU/512TQgQ0axh+n4N6itLwWt1i/sVts0eQTv4clCSP+9IDd4G69up1/PFJCs/DNPp3FovzSd36VfPgEE4cu+D4ziA6LoZNQjlIM6AnDTuIq06r1SUx/6Ldnv7c=",
            "0xDD579AF829306D32AC17A48b92e6111376160E7F":"AmuYcBgTKzmkaZj3njWLJdhHhFh/twXhERPwqqkl5jB5AotJRpNcaCofEFe2hS1VtKcBGHWuxojwYPlJXWiiprSpXRvyefKLfqgTXPB2ei9pHJca9dogdmiQvFtxwvHb0fMCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDWnQTDrnXukgZwyuK7bMuYWVuqpg722MmmWS/+bsv5lcC7Fx6saDhtZbXahl7L3eMq5vSuyQ8sCv4RZc1J6zkdl8DFM5QuMKzaDoF49QUqm/G7YbYJK+JbsHn9dP6Rhgkdl4CVJza+aEY4nzecsCHfiL9m1hzE/9PVEYVy23SUc8BU2VcenSCZ0TEuiSh2rfrM/r3O/niPoRbVfTv+5MhrtQte3bsDQaJnLIZbtRDKHafXAqQLpx2sdTf3YT5racl/6aMaA9Bi8NknqYakPMotYPSan59Icm0u1f7sS8lR9PZx9o="
         },
         "errors":{
            "0xfC3f3785e9Da47c6bb87889A5608DA2B55F51626":"Retrieval Kit 1 - Error Message 0"
         }
      },
      {
         "cfrags":{
            "0x91d3c58Fe14e04585B8Af6196A9D461CBa034524":"Az3OD0vu6K4p/Hn7Qat2xp+URZW6M53eQoKH/H7FGxkhAmlwOJGOica2g53WZWlekU45lHBkb6nebBAUGaHmikeK9/rH87xG3ne0eSPb9EbouwECYbGUz9nGPS2gVg6zs+kCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoCvz11Hhybksoa5ZyA65LTkYnzoB4Z92Se2LWbaj/rxr8DaS93NncmaiHaNifsfY0utF8WDi7qolIwcIOC+szOxsMDDLyPUhXP+QHTSl1GYw0BbWwM9eRHMrCDYRyInZuSUgMD82g7GIaOw6bsPUjwFjYxVG7+gD1/x/qeZnatsMurR5Pj/IcAWJBIrup6TDZolh5Nfb1hoSPLKQMucHaFiJIcJeErMAUp177vEpqIcZcPA0uzqOHDrutIcq6Rv1ulAYSPMAORkHnMNYrKVNkEFn9a4UcZ2C88gutVeu0B7rb1FTs=",
            "0x9cB2a48e029b210F7454ec268d65Cf2B7F0914f8":"AmlWmKHpCNC0B3cgwdUohuT2Et/ScbcFQIdJiGvrI1O1A75CljlW5dzC5qgM2KMDD0NWNWMJi/zaEoleoNj9LdxRLuj/LkGtMphva+2Y7RxIDffDcL98ko7xDoFpleIpYC0Ca4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDChbkY2wcAhKPVUlZ9t33cTehCOZHXmxYP1l+vQOFEZgCgdMlKvzgMV81c9GTA5DILza6V1bNqNQ1rzGB3KAUtgICvHEhOv4dhB4ABP+1liPLiNxs1ZsrUQm2ZFMhmVwl/o4DKFHpeO3JrhPfdRy85azu5IjmQ3z7F6ZCbxbvcTFRfye481X7B+8CtZGoX3/XpyOdeAtuSfQHAQzmac5qHYrAmIclCSP+9IDd4G69up1/PFJCs/DNPp3FovzSd36VfPgEE4cu+D4ziA6LoZNQjlIM6AnDTuIq06r1SUx/6Ldnv7c=",
            "0xDD579AF829306D32AC17A48b92e6111376160E7F":"AyJTNrXSDY+E5yvjNWWo9sv4MnOaGHG6pySTZ4gKBge+AykrmBVWOm4V766KHblrthIgKn7ErFBYhjUCOA3imuaxXRvyefKLfqgTXPB2ei9pHJca9dogdmiQvFtxwvHb0fMCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoCfaNSQYBtgg4LvOlyJYesMF5FripUyGwAcRXuDINqqJYCpWty/TNO7oQeB6rX5EQ6eeCuklRmiCrrHQ7wZTiKBuMDFM5QuMKzaDoF49QUqm/G7YbYJK+JbsHn9dP6Rhgkdl4DOz1Ag8amLVcQ7HF9w4X0zs3dQPdF23aAcnWW6cAz4Gvh7EqNKyYiOV18GwBj2xKRyxeI5HABhWY0sVXjrTkBTHbsDQaJnLIZbtRDKHafXAqQLpx2sdTf3YT5racl/6aMaA9Bi8NknqYakPMotYPSan59Icm0u1f7sS8lR9PZx9o="
         },
         "errors":{
            "0xa9267515f1cd2dBC89ED823bc3Eeddd6aD14C827":"Retrieval Kit 2 - Error Message 0",
            "0x2088348a8dF1E5B4F501f495C999f2A827825e39":"Retrieval Kit 2 - Error Message 1"
         }
      },
      {
         "cfrags":{
            "0x91d3c58Fe14e04585B8Af6196A9D461CBa034524":"A580rdo+LRzVjQCA79/WI6sgC6GTcTZWgQF3o5T/DL+jAz1jgeJwltMOQ08YxO7N8oB6Jl87x8h9r2Ad91sPu9Eb9/rH87xG3ne0eSPb9EbouwECYbGUz9nGPS2gVg6zs+kCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDqEbav8OGNY6G2g7tHK/uqgp40s3wVlSYel/5rZJzlIgCrfbR5oej4/RoosG3KTwB8kA3IPFp+80ygIpXtKlzvdoDDLyPUhXP+QHTSl1GYw0BbWwM9eRHMrCDYRyInZuSUgMCCh4IaJnEJoZC98eKLAMx1AuO4WOGmnh2hKLeJRWGzDjAFaXp+MviOS719Q6tIySeemg6+uOtDpUZIbzz9wSKieErMAUp177vEpqIcZcPA0uzqOHDrutIcq6Rv1ulAYSPMAORkHnMNYrKVNkEFn9a4UcZ2C88gutVeu0B7rb1FTs=",
            "0x9cB2a48e029b210F7454ec268d65Cf2B7F0914f8":"A3vtundidT5G5v5j7uTFXVZfITshAuYGmxtCHHJhT0FmA7kIVnz9AZf6myboE3W64Kkfx4kYPfwlZu5OIPE+TyLeLuj/LkGtMphva+2Y7RxIDffDcL98ko7xDoFpleIpYC0Ca4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoDrwmxdPXYVr/usAe7qLhQMy8TpmfJxpSR9k1RKq/5tvYCt9+JGKq8pfCVVNsW3YxTy/lJ5crrCUW/Zqrq3Ntba1kCvHEhOv4dhB4ABP+1liPLiNxs1ZsrUQm2ZFMhmVwl/o4DzAlZVtF4n8x/xgEaAnFZxICbRdVr1cpq0vLx6DpyGnZJHK/3bgNIV824CynQohOcvB4tW1R3gjF7H8V4cyY3K4clCSP+9IDd4G69up1/PFJCs/DNPp3FovzSd36VfPgEE4cu+D4ziA6LoZNQjlIM6AnDTuIq06r1SUx/6Ldnv7c=",
            "0xDD579AF829306D32AC17A48b92e6111376160E7F":"A5KHFYVdxW6RVuMt1au5f9Gg2UbY2FMoiCSGnrRAOCHUAyfMbKe8qvD04CrVN3jAh2pBJ8YwS2yZzcNGnkoa1g0PXRvyefKLfqgTXPB2ei9pHJca9dogdmiQvFtxwvHb0fMCa4WvOWCW0XtsnJvjkeSr1tRuFi9Gm83a7yf+i1XyLPoCgT3p+y6yvj/fE2c7SNfyuOg15odD6jaiGEmTz0n9+HQDulcxswVh8/Dqge9QW/J+iu0vS2qakrNgaTq4pewCsioDFM5QuMKzaDoF49QUqm/G7YbYJK+JbsHn9dP6Rhgkdl4Dgy3fzwcpzP56tybN29woXcaxvlxhHcCENN0j8tF2A/rEPo7hFsycMcYbh6QmDU4YZbLm5bjn6KVwPRpB/U5L53bsDQaJnLIZbtRDKHafXAqQLpx2sdTf3YT5racl/6aMaA9Bi8NknqYakPMotYPSan59Icm0u1f7sS8lR9PZx9o="
         },
         "errors":{
            "0xbF4C25CB226CE41864d8807F0202307E1A8CDe74":"Retrieval Kit 3 - Error Message 0",
            "0x11a2cC215852Dd09188cEf5842F8AC12a50F207b":"Retrieval Kit 3 - Error Message 1",
            "0x9962B21C211f3192CEc88af27D3Ca816d7DC6525":"Retrieval Kit 3 - Error Message 2"
         }
      }
   ]
}
```